### ISSUE TITLE
fill may render on wrong side of incomplete multipolygons

### ISSUE THREAD

**Issue by tmcw** [openstreetmap/iD#1201]
![](http://dl.dropbox.com/u/68059/Screenshots/_7wbp9r1lkyu.png)

[This way/relation has the tag waterway=polygon and renders incorrectly](http://www.openstreetmap.org/browse/relation/2389611)

**Comment by slhh** [comment]
The rendering issue of broken or incompletely loaded multipolygons could be addressed in the following way: Define additional roles like outer:right and outer:left indicating on which side of the way the multipolygon area is located. The user should not need to add these suffices, but the editor can automatically add or update these, if the multipolygon is complete. This would significantly

Being a change of the OSM datastructures this needs to be discussed with the community in case the change is desired.

**Comment by bhousel** [comment]
I guess much has been written about this issue since 2013.  see: The [Future of Areas](http://wiki.openstreetmap.org/wiki/Area/The_Future_of_Areas) for a current summary. 

So where things stand right now, iD will always wrongly render certain multipolygons unless we download the entire multipolygon.  Some multipolygons are too prohibitively large to download.

@slhh's solution of adding tags to indicate which side of the way contains the area could work.  (It sounds a bit like [Tagging Outlines](http://wiki.openstreetmap.org/wiki/The_Future_of_Areas/Tagging_Outlines) from the above link, but not exactly).   It would definitely upset a lot of people in the OSM community if iD started silently tagging multipolygon member ways with this extra data.

That said, we can use [Any Tags We Like](http://wiki.openstreetmap.org/wiki/Any_tags_you_like).  I'm kind of inclined to push OSM in this direction and just start doing it.  Thoughts @pnorman?

**Comment by pnorman** [comment]
> That said, we can use Any Tags We Like. I'm kind of inclined to push OSM in this direction and just start doing it. Thoughts @pnorman?

Don't.

> So where things stand right now, iD will always wrongly render certain multipolygons unless we download the entire multipolygon. Some multipolygons are too prohibitively large to download.

I haven't found this an issue in practice. Do you have data which indicates otherwise?

I'd still recommend `/relation/id#/full`. I can probably get some solid data on it, but my instinct is that most `type=multipolygon` relations will be entirely within the downloaded area, and of those that aren't, most will be small.

**Comment by bhousel** [comment]
> I haven't found this an issue in practice. Do you have data which indicates otherwise?
> 
> I'd still recommend /relation/id#/full. I can probably get some solid data on it, but my instinct is that most type=multipolygon relations will be entirely within the downloaded area, and of those that aren't, most will be small.

@pnorman I went back and read [your comment](https://github.com/openstreetmap/iD/issues/1201#issuecomment-24951533) a bit more carefully. 

I originally thought we would need to do this background `relation/id/full` fetch for all relations, but if we limit it to only multipolygons it would probably be ok.  Coastlines and boundaries are where this approach wouldn't scale.

**Comment by pnorman** [comment]
> Coastlines and boundaries are where this approach wouldn't scale.

Yes, boundaries could be large, and take several seconds. Coastlines don't have relations.

I did some numbers, and of the 2 million MP relations

```
members   number
1           86k
2         1000k
3          260k
4          230k
5          101k
>5         362k
```

**Comment by bryceco** [comment]
The total number of nodes in the MP is more interesting to me than the number of members. That's a better measure of bandwidth and memory cost.

**Comment by bhousel** [comment]
> The total number of nodes in the MP is more interesting to me than the number of members. That's a better measure of bandwidth and memory cost.

Yes, but that number is capped at 2000 nodes per member way.   It's probably ok.

I don't have time this week, but sometime soon I'll try automatically fetching the multipolygons with `relation/id/full`.  My home state of New Jersey is a great place to test this - it is covered with huge imported landuse multipolygons.

**Comment by bhousel** [comment]
So, @pnorman maybe the more interesting question to ask is how many members these multipolygons max out at.   ">5" doesn't sound so scary, but if the max is thousands then we have a problem.

Here's an outlier mentioned in IRC:  https://www.openstreetmap.org/relation/5989495
We should test how long that takes to download via `relation/id/full`

**Comment by emil813-cmd** [comment]
Theoretically, if multipolygon used **forward/backward** instead of inner/outer, then this problem wouldn't exist. What's on the "left" side would be inside the multipolygon, and what's on the "right" side would be outside.

But this would require all apps to adapt to this system. It would also make multipolygons really annoying to create, unless the editor converts them to inner/outer while you are editing the relation.

EDIT: I guess this problem is beyond the scope of the iD editor.
